internal/pkg/vm/internal/vmcontext/vmcontex2.go

```
@@ -0,0 +1,543 @@
+		vm.applyImplicitMessage(epostMessage)
+
+		// initial miner penalty
+		// Note: certain msg execution failures can cause the miner to pay for the gas
+		minerPenaltyTotal := types.ZeroAttoFIL
+
+		// Process BLS messages from the block
+		for _, m := range blk.BLSMessages {
+			// do not recompute already seen messages
+			mcid := msgCID(m)
+			if _, found := seenMsgs[mcid]; found {
+				continue
+			}
+
+			// apply message
+			receipt, minerPenaltyCurr := vm.applyMessage(m, m.OnChainLen(), blk.Miner)
+
+			// accumulate result
+			minerPenaltyTotal = minerPenaltyTotal.Add(minerPenaltyCurr)
+			receipts = append(receipts, receipt)
+
+			// flag msg as seen
+			seenMsgs[mcid] = struct{}{}
+		}
+
+		// Process SECP messages from the block
+		for _, sm := range blk.SECPMessages {
+			// extract unsigned message part
+			m := sm.Message
+
+			// do not recompute already seen messages
+			mcid := msgCID(&m)
+			if _, found := seenMsgs[mcid]; found {
+				continue
+			}
+
+			// apply message
+			// Note: the on-chain size for SECP messages is different
+			receipt, minerPenaltyCurr := vm.applyMessage(&m, sm.OnChainLen(), blk.Miner)
+
+			// accumulate result
+			minerPenaltyTotal = minerPenaltyTotal.Add(minerPenaltyCurr)
+			receipts = append(receipts, receipt)
+
+			// flag msg as seen
+			seenMsgs[mcid] = struct{}{}
+		}
+
+		// Pay block reward.
+		rewardMessage := makeBlockRewardMessage(blk.Miner, minerPenaltyTotal)
+		vm.applyImplicitMessage(rewardMessage)
+	}
+
+	// cron tick
+	firstMiner := blocks[0].Miner
+	cronMessage := makeCronTickMessage(firstMiner)
+	vm.applyImplicitMessage(cronMessage)
+
+	// commit state
+	// flush all objects out
+	if err := vm.store.Flush(); err != nil {
+		// Review: what do we want to do with this errors?
+		//         we can change the return to be ([]Receipt, error) although its unclear what to do about it as a caller
+		//         we can just crash, or let the caller retry a few times and then crash
```

Here and below I think we should return an error value and propagate it up.